src/Collection/CollectionTrait.php

```diff
@@ -692,23 +692,38 @@ public function cartesianProduct(callable $operation = null, callable $filter =
             return new Collection([]);
         }
 
+        // contains each array of the collection
+        $values = [];
+        // contains the keys of each array
+        $valuesKeys = [];
+        // contains the number of elements of each array
+        $counts = [];
+
+        foreach ($this as $value) {
+            // fail if any of the values is a multidimensional array
```

Comment:
I dont think all the inline comments are necessary, with a good exception etc here
also the ones above should be clear from the variable names ideally.